ros2_generate_interface_docs/ros2_generate_interface_docs/utils.py

```diff
@@ -0,0 +1,451 @@
+# Copyright 2020 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import array
+import errno
+import html
+from io import StringIO
+import os
+import shutil
+import sys
+
+import em
+
+from rosidl_parser.definition import AbstractNestedType, BoundedSequence, NamespacedType
+
+from rosidl_runtime_py import get_interface_path
+
+
+_TEMPLATES_DIR = 'templates'
+IGNORED_KEYS = [
+    '__slots__', '__doc__', '_fields_and_field_types', 'SLOT_TYPES', '__init__',
+    '__repr__', '__eq__', '__hash__', 'get_fields_and_field_types', '__module__'
+]
+
+BASIC_TYPE_CONVERSION = {
+    'boolean': 'bool',
+    'octet': 'bytes',
+    'uint8': 'uint8',
+    'int8': 'int8',
+    'uint16': 'uint16',
+    'int16': 'int16',
+    'uint32': 'uint32',
+    'int32': 'int32',
+    'uint64': 'uint64',
+    'int64': 'int64',
+    'char': 'str',
+    'float': 'float32',
+    'double': 'float64',
+    'short': 'int',
+    'unsigned short': 'int',
+    'long': 'int',
+    'unsigned long': 'int',
+    'long long': 'int',
+    'unsigned long long': 'int',
+    'string': 'string',
+    'wstring': 'wstring'
+}
```

Comment:
What is this map used for? Which language is it for? The language specific code generator probably also defines this mapping.

Reply:
When I call the method `get_fields_and_field_types` for a specific interface this will return the Python type of the fields. this step is needed to get the type of the field and check if this a basic type, array namespace or nested. This dictionary converts from Python types to interfaces types

In `rosidl_generator_py` you can find the conversion from interface types to Python types (listed below) but not opposite way:

 - `get_python_type()`
 - `constant_value_to_py()`
 - `primitive_value_to_py()`

should I need to create this method? 

 - from_py_to_interface_type()

or it's okay with the dictionary?

Reply:
Using `parse_idl_file()` https://github.com/ros2/ros2_generate_interface_docs/pull/4/commits/68c54259e21bb4ca496f82387759e746d58d31a6